MediaBrowser.Providers/Trickplay/TrickplayProvider.cs

```diff
@@ -0,0 +1,126 @@
+    public TrickplayProvider(
+        ILogger<TrickplayProvider> logger,
+        IServerConfigurationManager config,
+        ITrickplayManager trickplayManager,
+        ILibraryManager libraryManager)
+    {
+        _logger = logger;
+        _config = config;
+        _trickplayManager = trickplayManager;
+        _libraryManager = libraryManager;
+    }
+
+    /// <inheritdoc />
+    public string Name => "Trickplay Provider";
+
+    /// <inheritdoc />
+    public int Order => 100;
+
+    /// <inheritdoc />
+    public bool HasChanged(BaseItem item, IDirectoryService directoryService)
+    {
+        if (item.IsFileProtocol)
+        {
+            var file = directoryService.GetFile(item.Path);
+            if (file is not null && item.DateModified != file.LastWriteTimeUtc)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <inheritdoc />
+    public Task<ItemUpdateType> FetchAsync(Episode item, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        return FetchInternal(item, options, cancellationToken);
+    }
+
+    /// <inheritdoc />
+    public Task<ItemUpdateType> FetchAsync(MusicVideo item, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        return FetchInternal(item, options, cancellationToken);
+    }
+
+    /// <inheritdoc />
+    public Task<ItemUpdateType> FetchAsync(Movie item, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        return FetchInternal(item, options, cancellationToken);
+    }
+
+    /// <inheritdoc />
+    public Task<ItemUpdateType> FetchAsync(Trailer item, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        return FetchInternal(item, options, cancellationToken);
+    }
+
+    /// <inheritdoc />
+    public Task<ItemUpdateType> FetchAsync(Video item, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        return FetchInternal(item, options, cancellationToken);
+    }
+
+    private async Task<ItemUpdateType> FetchInternal(Video video, MetadataRefreshOptions options, CancellationToken cancellationToken)
+    {
+        var libraryOptions = _libraryManager.GetLibraryOptions(video);
+        bool? enableDuringScan = libraryOptions?.ExtractTrickplayImagesDuringLibraryScan;
+        bool replace = options.ReplaceAllImages;
+
+        if (options.IsAutomated && !enableDuringScan.GetValueOrDefault(false))
+        {
+            return ItemUpdateType.None;
+        }
+
+        if (_config.Configuration.TrickplayOptions.ScanBehavior == TrickplayScanBehavior.Blocking)
+        {
+            await _trickplayManager.RefreshTrickplayDataAsync(video, replace, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            _ = _trickplayManager.RefreshTrickplayDataAsync(video, replace, cancellationToken).ConfigureAwait(false);
```

Comment:
I'm not a fan of having a nonblocking call for this

Reply:
Blocking the library scan was a very unpopular choice that caused a ton of confusion which is why a non-blocking option was added and made the default. Plus, trickplay data is non-critical and not really necessary information when it comes to adding something to the library, so blocking the scan for upwards of 10 minutes per video depending on hardware would mean almost nobody would use this feature.

Reply:
I believe this is how the code currently functions. Users can selectively enable trickplay extraction and extraction on scan for each individual library. Only libraries with extraction enabled will ever have trickplay images generated, but libraries with trickplay enabled and scan generation disabled will have images generated through the scheduled task.

The enabled library check, along with other checks like media stream count, is done in the TrickplayManager to avoid code duplication across the task and metadata provider. The TrickplayProvider itself checks for enable on scan since that's the only code run on library scan. 

https://github.com/jellyfin/jellyfin/blob/f1874e2a489e0ed042021cefcf7b3d008ea3bf33/MediaBrowser.Providers/Trickplay/TrickplayManager.cs#L319-L323 

https://github.com/jellyfin/jellyfin/blob/f1874e2a489e0ed042021cefcf7b3d008ea3bf33/MediaBrowser.Providers/Trickplay/TrickplayProvider.cs#L101-L102

Reply:
Ah perfect, I missed that call in `CanGenerateTrickplay`

I'm still not a fan of the nonblocking call, but I guess it can stay for now